fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/JoinedRecordTypeBuilder.java

```diff
@@ -73,17 +74,46 @@ public static class Join {
         @Nonnull
         private final String left;
         @Nonnull
-        private final KeyExpression leftExpression;
+        private final KeyExpression leftIndexExpression;
+        @Nonnull
+        private final KeyExpression leftJoinExpression;
         @Nonnull
         private final String right;
         @Nonnull
-        private final KeyExpression rightExpression;
+        private final KeyExpression rightIndexExpression;
+        @Nonnull
+        private final KeyExpression rightJoinExpression;
 
-        public Join(@Nonnull String left, @Nonnull KeyExpression leftExpression, @Nonnull String right, @Nonnull KeyExpression rightExpression) {
+        /**
+         * Create a new join from left to right.
+         *
+         * @param left the name of the type on the LHS of the join.
+         * @param leftIndexExpression the KeyExpression to use when constructing indexes. The entries in this
+         * KeyExpression
+         * must line up correctly with how the join index will store records on disk.
```

Comment:
This doesn't seem to quite capture it. In particular, there is. no _requirement_ that these keys' values be stored in the index at all. Often they are the primary key on one side, which is stored, but nothing prevents joining on some kind of surrogate instead.

I'm trying to think what the most succinct way to describe what's going on it, without introducing a bunch of higher level machinery down here in the model. Perhaps the missing concept is the _synthesis_. Given one of the (assume two, but it extends to n) constituents, how to we find the other to build the synthetic joined record? It's actually as simple as could be: take an _index_ key expression from the record we have and evaluate it into a query binding. Then build a query where the _join_ key expression from the other side equals that binding. The results of this query are the other records we need to synthesize.

All indexing (which is not restricted by index type -- "join indexes" are a classification of the derivation, not a type of index) happens on the synthetic records, which have nested fields for each component and a compound primary key.

Reply:
I worked out an example case and tried to explain what was going on that way.